[mlir][ExecutionEngine] Add LevelZeroRuntimeWrapper.

mshahneo opened the pull request:
Adds LevelZeroRuntime wrapper and tests.

Co-authored-by: Artem Kroviakov <artem.kroviakov@intel.com>
Co-authored-by: Nishant Patel <nishant.b.patel@intel.com>

mshahneo committed:
[mlir][ExecutionEngine] Add LevelZeroRuntimeWrapper.

Adds LevelZeroRuntime wrapper and tests.

Co-authored-by: Artem Kroviakov <artem.kroviakov@intel.com>
Co-authored-by: Nishant Patel <nishant.b.patel@intel.com>

Hardcode84 commented:
@mshahneo we already have `SYCLRuntimeWrappers`, why not use those? Also, other folks from Intel are working on offload runtime upstream, can you please sync with them if their work can be reused? I would very much prefer not to multiply these terrible wrappers.

mshahneo committed:
Fix CMake options.

rengolin commented on mlir/lib/ExecutionEngine/CMakeLists.txt:
```diff
@@ -100,9 +101,9 @@ mlir_target_link_libraries(MLIRExecutionEngine PUBLIC
+)
 
-if(LLVM_BUILD_LLVM_DYLIB AND NOT (WIN32 OR MINGW OR CYGWIN)) # Does not build on windows currently, see #106859
+if(LLVM_BUILD_LLVM_DYLIB AND NOT(WIN32 OR MINGW OR CYGWIN)) # Does not build on windows currently, see #106859
```
Please avoid spurious whitespace changes. These increase greatly the amount of code to review when they should not have changed.

mshahneo replied:
Hi Renato (@rengolin ),

I apologize for this, I had a CMake formatter in my vscode which did auto-formatting :(.
Should I remove the formatted version?

rengolin replied:
At least the parts that do not change otherwise. I'd recommend you disable auto-format in CMake files and only run on the lines you change and surrounding code. (Most editors can select code and run the formatter on the selection).

mshahneo replied:
Fixed now. Thanks, @rengolin :).

rengolin reviewed (commented):
I cannot comment on the L0 wrapper logic. @jian-lin @kurapov-peter please have a look.

@Hardcode84 the idea is to move away from the SYCL runner into a pure L0 runner so that we don't need to depend on the SYCL runtime when we're not using it. Once L0 is the default way to run on Intel GPUs, we should remove the SYCL runtime logic. SYCL can also use the L0 runtime.

@mshahneo, there are too many unrelated whitespace changes in this PR. Please roll-back any formatting change that you have done and only keep the actual logic changes in this PR. Thanks!

adam-smnk commented on mlir/test/Integration/GPU/LEVELZERO/gpu-addf32-to-spirv.mlir:
nit: for consistency the directory name should also use camel case for level zero

adam-smnk replied:
The test file names could also drop `gpu` and `to-spirv` for less noise. Unless there's a plan to have some other variants as well.

mshahneo replied:
Actually the test folders in this tree all are named all caps. That's why I used it. Keeping `gpu` and `to-spirv` for future in mind.

adam-smnk replied:
I just happens that most of these are acronyms/names that use caps in their branding. Except `Vulkan` and I don't think Level Zero goes for all caps normally.

mshahneo replied:
Fixed.

adam-smnk commented on mlir/test/Integration/GPU/LEVELZERO/gpu-addf32-to-spirv.mlir:
```diff
@@ -0,0 +1,56 @@
+    %memref_0 = gpu.alloc host_shared () : memref<2x2x2xf32>
+    memref.copy %arg0, %memref_0 : memref<2x2x2xf32> to memref<2x2x2xf32>
+    %memref_2 = gpu.alloc host_shared () : memref<2x2x2xf32>
+    %2 = gpu.wait async
```
Is `async` really needed?

adam-smnk commented on mlir/test/Integration/GPU/LEVELZERO/gpu-addf32-to-spirv.mlir:
```diff
@@ -0,0 +1,56 @@
+    memref.copy %arg0, %memref_0 : memref<2x2x2xf32> to memref<2x2x2xf32>
+    %memref_2 = gpu.alloc host_shared () : memref<2x2x2xf32>
+    %2 = gpu.wait async
+    %3 = gpu.launch_func async [%2] @test_kernel::@test_kernel blocks in (%c2, %c2, %c2) threads in (%c1, %c1, %c1) args(%memref_0 : memref<2x2x2xf32>, %mem : memref<2x2x2xf32>, %memref_2 : memref<2x2x2xf32>)
```
nit: line breaks for readability

mshahneo replied:
I see, got it.
Initially I didn't touch it because I saw large lines in .mlir code.
Fixed now.

adam-smnk commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+#include <exception>
+#include <functional>
+#include <iostream>
+#include <level_zero/ze_api.h>
```
nit: maybe use `" "` for the include, it's not really a standard library

mshahneo replied:
Thanks, done :)

joker-eph commented on mlir/test/Integration/GPU/LEVELZERO/gpu-reluf32-to-spirv.mlir:
```diff
@@ -0,0 +1,79 @@
+// RUN: mlir-opt %s -pass-pipeline='builtin.module(spirv-attach-target{ver=v1.0 caps=Addresses,Int64,Kernel},convert-gpu-to-spirv{use-64bit-index=true},gpu.module(spirv.module(spirv-lower-abi-attrs,spirv-update-vce)),func.func(llvm-request-c-wrappers),convert-scf-to-cf,convert-cf-to-llvm,convert-arith-to-llvm,convert-math-to-llvm,convert-func-to-llvm,gpu-to-llvm{use-bare-pointers-for-kernels=true ...
```
Can you use `convert-to-llvm`? That should be the modern way to perform LLVM conversion:

```suggestion
// RUN: mlir-opt %s -pass-pipeline='builtin.module(spirv-attach-target{ver=v1.0 caps=Addresses,Int64,Kernel},convert-gpu-to-spirv{use-64bit-index=true},gpu.module(spirv.module(spirv-lower-abi-attrs,spirv-update-vce)),func.func(llvm-request-c-wrappers),convert-scf-to-cf,convert-to-llvm,gpu-module-to-binary,expand-strided-metadata,lower-affine,finalize-memref-to-llvm,reconcile-unrealized-casts)' \
```

I see that there is an option for `gpu-to-llvm{use-bare-pointers-for-kernels=true}` so that may have to be done ahead of time separately, but we should look into using the dlti infra to replace this in the future I think (@fabianmcg : is this on your radar maybe?)

mshahneo replied:
Thanks, @joker-eph. Using `-convert-to-llvm` for all `to-llvm` conversions except `-gpu-to-llvm`.
I think we should be able to remove it once @fabianmcg 's changes land.

joker-eph commented on mlir/test/Integration/GPU/LEVELZERO/gpu-reluf32-to-spirv.mlir:
```diff
@@ -0,0 +1,79 @@
+// RUN:   --shared-libs=%mlir_sycl_runtime \
+// RUN:   --shared-libs=%mlir_runner_utils \
+// RUN:   --entry-point-result=void \
+// RUN: | FileCheck %s
```
I'm confused, this test does not use the mlir_levelzero_runtime? Why does it relate to this PR?

mshahneo committed:
Address review comments.

Hardcode84 commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+  }
+
+  void enqueueOp(
+      std::function<void(ze_event_handle_t, uint32_t, ze_event_handle_t *)>
```
`std::function` objects can be expensive to create, use `llvm::function_ref` or templated functor
```
template<typenamr F>
enqueueOp(F&& op)
```

mshahneo replied:
Thanks, done.

Hardcode84 commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+}
+
+extern "C" void mgpuStreamDestroy(StreamWrapper *stream) {
+  if (stream)
```
`delete nullptr` is allowed, this check is redundant.

mshahneo replied:
Done.

Hardcode84 commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+
+extern "C" void mgpuStreamWaitEvent(StreamWrapper *stream,
+                                    ze_event_handle_t event) {
+  assert(stream && event);
```
```
assert(stream && "Invalid stream");
assert(event && "Invalid stream");
```

Hardcode84 commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+void mgpuMemset(void *dst, PATTERN_TYPE value, size_t count,
+                StreamWrapper *stream) {
+  auto listType =
+      getRtContext().copyEngineMaxMemoryFillPatternSize >= sizeof(PATTERN_TYPE)
```
call `getRtContext()` once and take the reference.

mshahneo replied:
Done.

Jianhui-Li commented:
> I cannot comment on the L0 wrapper logic. @jian-lin @kurapov-peter please have a look.
> 
> @Hardcode84 the idea is to move away from the SYCL runner into a pure L0 runner so that we don't need to depend on the SYCL runtime when we're not using it. Once L0 is the default way to run on Intel GPUs, we should remove the SYCL runtime logic. SYCL can also use the L0 runtime.

The L0 runner removes the dependencies on SYCL runtime and also header files shipped together with SYCL compiler, so it simplifies MLIR development on Intel GPU. 

SYCL runtime is not only for Intel GPU. It targets all SYCL compatible devices. We are open to remove it if there are no interests from the MLIR community to use it.

Jianhui-Li commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+
+  thread_local static bool isDriverInitialised{false};
+  if (isDriverInitialised)
+    return drivers[idx];
```
Do you need to compare idx with the size of drivers? I mean, what if the driver is initialized but the drivercount is less than idx requested?

mshahneo replied:
Sorry, thanks for the catch, fixed now.

Jianhui-Li commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+                                           const int32_t devIdx = 0) {
+  thread_local static ze_device_handle_t l0Device;
+  thread_local static int32_t currDevIdx{-1};
+  if (devIdx == currDevIdx)
```
Is it possible that user give different driverIdx but same devIdx? In this case, they get same l0Device?

mshahneo replied:
Fixed.

Jianhui-Li commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+};
+
+// L0 only supports pre-determined sizes of event pools,
+// implement a rt data struct to avoid running out of events.
```
nit: rt-> runtime

Jianhui-Li commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+    eventPools.push_back(nullptr);
+    L0_SAFE_CALL(zeEventPoolCreate(rtCtx->context, &eventPoolDesc, 1,
+                                   &rtCtx->device, &eventPools.back()));
+    currentEventsLimit += numEvents;
```
What happens if the number of even and event pool surpass the "pre-determined sizes of event pools" supported by L0?

Jianhui-Li replied:
I think a limit could be useful - just to be defensive in case something goes wrong.

mshahneo replied:
Added `maxEventsCount` which is set to 32K events. It should be enough.

Jianhui-Li commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+  }
+
+  ~DynamicEventPool() {
+    assert(!takenEvents.size());
```
consider using std::unique_ptr with custom deletes, so you don't have to assert these conditions.  Same comments as above https://github.com/llvm/llvm-project/pull/151038/files#r2240787050

mshahneo committed:
Use std::unique_ptr for L0 handles.

Fix Cmake whitespace isseu.

kurapov-peter commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+                                              &immCmdListCompute));
+  }
+  void cleanup() {
+    L0_SAFE_CALL(zeCommandListDestroy(immCmdListCopy));
```
+1, and if you leave it as is, there's potential for inconsistent states (e.g., user called `cleanup`, then `getDefaultContext` and got an invalid context handle.

mshahneo replied:
Thanks, @Hardcode84 , @kurapov-peter!
Changed to `std::unique_ptr`

kurapov-peter commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+    ze_event_pool_desc_t eventPoolDesc = {};
+    eventPoolDesc.flags = ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
+    eventPoolDesc.count = numEvents;
+    eventPools.push_back(nullptr);
```
this looks a bit confusing

mshahneo committed:
Use std::unique_ptr for DynamicEventPool members.

mshahneo committed:
Address review comments.

Add L0 error message.
Find levelzero only once if SYCL or LEVELZERO runner is enabled.

mshahneo committed:
Address review comments.

Add linebreak for large lines.
Rename test directory to 'LevelZero'.

mshahneo committed:
Address review comments.

Revert formatting added to Cmake File.

mshahneo committed:
Fix end-of-line issue.

mshahneo commented on mlir/cmake/modules/FindLevelZeroRuntime.cmake:
```diff
@@ -20,40 +20,39 @@ include(FindPackageHandleStandardArgs)
     )
 
-    find_library(LevelZero_LIBRARY
+    find_library(LevelZeroRuntime_LIBRARY
```
I did it to make it consistent with SYCL. Since it's a one time thing, planning to keep it if no opposition.

mshahneo commented on mlir/lib/ExecutionEngine/CMakeLists.txt:
```diff
@@ -395,18 +410,51 @@ if(LLVM_ENABLE_PIC)
+  endif()
+
+  if(MLIR_ENABLE_LEVELZERO_RUNNER)
+    find_package(LevelZeroRuntime)
```
Yes, thanks, fixed it now.

mshahneo commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,491 @@
+  {                                                                            \
+    ze_result_t status = (call);                                               \
+    if (status != ZE_RESULT_SUCCESS) {                                         \
+      std::cerr << "L0 error " << status << std::endl;                         \
```
Done.

Jianhui-Li commented on mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp:
```diff
@@ -0,0 +1,567 @@
+using UniqueZeCommandList =
+    std::unique_ptr<std::remove_pointer<ze_command_list_handle_t>::type,
+                    ZeCommandListDeleter>;
+struct L0RtContext {
```
nit: consider L0RtContextWrapper？to differentiate with the L0 context returned from zeContextCreate

mshahneo committed:
Address review comments.

Limit the maximum no. of events.
Modify getContext to be a bit more generic.

Jianhui-Li reviewed (approved):
LGTM

adam-smnk commented on mlir/lib/ExecutionEngine/CMakeLists.txt:
```diff
@@ -404,9 +408,33 @@ if(LLVM_ENABLE_PIC)
+      EXCLUDE_FROM_LIBMLIR
+    )
+
+    # check_cxx_compiler_flag("-frtti" CXX_HAS_FRTTI_FLAG)
+
+    # if(NOT CXX_HAS_FRTTI_FLAG)
+    # message(FATAL_ERROR "CXX compiler does not accept flag -frtti")
+    # endif()
```
Remove unused code

mshahneo committed:
Address review comments.

Remove unused code from CMake file.

mshahneo commented:
@joker-eph , @rengolin , @fabianmcg , @kurapov-peter, any more comments?

mshahneo committed:
Remove new line changes in Cmake File.
